### PR TITLE
Jankbench collector

### PR DESCRIPTION
**Pull request by credp**
Here's my WIP Jankbench collector, plus a couple of notebooks showing how they can be used.

### CODE REVIEW

File: lisa/wa.py
Comment (credp) [inline lisa/wa.py:576]: I noticed that this ends up throwing an error when we open the collector and the results set contains non-jankbench workloads. I probably only want an error if the workload should be expected to have jankbench results present and we failed to open then...

What would be the recommended way to deal with that - return None from _get_artifact_df and throw for errors in loading?

Comment (douglas-raillard-arm) [reply lisa/wa.py:576]: > I noticed that this ends up throwing an error when we open the collector and the results set contains non-jankbench workloads

In `WAArtifactCollectorBase`:
```
    def _get_job_df(self, job):
        path = job.get_artifact_path(self._ARTIFACT_NAME)
        df = self._get_artifact_df(path)
        return self._add_job_info(job, df)
```
Do you mean that some jobs may not have run Jankbench, but we end up querying `job.get_artifact_path(self._ARTIFACT_NAME)` on it, which returns something unexpected (None?) and makes your method fail ?

Comment (douglas-raillard-arm) [reply lisa/wa.py:576]: From what I can see, we already handle any exception raised in `_get_job_df()` in `_get_df()`:
```
            try:
                df = loader(job)
            except Exception as e:
                self.logger.error('Could not load {} dataframe for job {}: {}'.format(self.NAME, job, e))
                return None
```
(`loader()` calls `self._get_job_df(job)`)
If there is a way to tell from a job that it should not be expected to have a given artifact, we could avoid having the logging.error().

Comment (douglas-raillard-arm) [reply lisa/wa.py:576]: Checkout this branch and I'll merge it if it solves the problem:
https://github.com/ARM-software/lisa/pull/1538

Comment (credp) [reply lisa/wa.py:576]: It looks good to me, I'll try it out and let you know

Comment (douglas-raillard-arm) [reply lisa/wa.py:576]: Fixed by: https://github.com/ARM-software/lisa/pull/1538

**Review by douglas-raillard-arm — CHANGES_REQUESTED** [review]
Had a look over the WAJankbenchCollector commit, I'll wait until the notebook becomes more stable to review it

File: lisa/wa.py
Comment (douglas-raillard-arm) [inline lisa/wa.py:578]: We might want to select the columns we are actually interested in, since we already have a hardcoded list of them anyway. This should make loading faster and less memory hungry

File: lisa/wa.py
Comment (credp) [reply lisa/wa.py:36]: Done, removed.

File: lisa/wa.py
Comment (credp) [reply lisa/wa.py:37]: Done!

File: lisa/wa.py
Comment (credp) [reply lisa/wa.py:557]: Good point, I'll remove all of that type stuff and preformat it,

File: lisa/wa.py
Comment (credp) [reply lisa/wa.py:578]: Yep, this is left over from when I wasn't so sure what I wanted to show. Fixed.

File: lisa/wa.py
Comment (credp) [reply lisa/wa.py:556]: Done, thanks for the pointer.

File: lisa/wa.py
Comment (credp) [reply lisa/wa.py:579]: I think I can do the rename in the SQL as well, so we only create one df and do it right in the beginning

File: lisa/wa.py
Comment (credp) [reply lisa/wa.py:593]: OK, I'm not sure I understand what you mean here. Let's have a chat so I can catch up!

File: lisa/wa.py
Comment (credp) [reply lisa/wa.py:606]: I think I've fixed this - it looks to me like it matches your request in the version I just pushed.

Comment (douglas-raillard-arm) [reply lisa/wa.py:606]: The code looks like it's now doing the right thing, would it be possible to paste here a sample of the output dataframe to confirm ?

File: lisa/wa.py
Comment (douglas-raillard-arm) [reply lisa/wa.py]: Actually I got confused, `sqlite3` is part of the standard library so it would go in the first group of imports, not the 2nd one, sorry about the churn :(

File: lisa/wa.py
Comment (credp) [reply lisa/wa.py:606]: frame_id | variable | value | iteration | workload | id | tag | kernel
-- | -- | -- | -- | -- | -- | -- | --
1 | total_duration | 27.068337 | 1 | jankbench | default-jankbench_list_view | iter_3 | 4.14.170-gc72771da3_audio-g3ab4335-ab32
2 | total_duration | 19.426523 | 1 | jankbench | default-jankbench_list_view | iter_3 | 4.14.170-gc72771da3_audio-g3ab4335-ab32
3 | total_duration | 4.794331 | 1 | jankbench | default-jankbench_list_view | iter_3 | 4.14.170-gc72771da3_audio-g3ab4335-ab32
4 | total_duration | 4.816895 | 1 | jankbench | default-jankbench_list_view | iter_3 | 4.14.170-gc72771da3_audio-g3ab4335-ab32
5 | total_duration | 4.369130 | 1 | jankbench | default-jankbench_list_view | iter_3 | 4.14.170-gc72771da3_audio-g3ab4335-ab32
... | ... | ... | ... | ... | ... | ... | ...
149 | jank_frame | 0.000000 | 3 | jankbench | default-jankbench_edit_text | iter_3 | 4.14.111_audio-g3ab4335-ab31
150 | jank_frame | 0.000000 | 3 | jankbench | default-jankbench_edit_text | iter_3 | 4.14.111_audio-g3ab4335-ab31
151 | jank_frame | 0.000000 | 3 | jankbench | default-jankbench_edit_text | iter_3 | 4.14.111_audio-g3ab4335-ab31
152 | jank_frame | 0.000000 | 3 | jankbench | default-jankbench_edit_text | iter_3 | 4.14.111_audio-g3ab4335-ab31
153 | jank_frame | 0.000000 | 3 | jankbench | default-jankbench_edit_text | iter_3 | 4.14.111_audio-g3ab4335-ab31

Comment (credp) [reply lisa/wa.py:606]: The stats output looks like:



id | kernel | tag | variable | workload | stat | value | unit | ci_minus | ci_plus
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
default-jankbench_edit_text | 4.14.111_audio-g3ab4335-ab31 | iter_3 | jank_frame | jankbench | median | NaN | % | NaN | NaN
default-jankbench_edit_text | 4.14.111_audio-g3ab4335-ab31 | iter_3 | jank_frame | jankbench | mean | -0.871460 | % | NaN | NaN
default-jankbench_edit_text | 4.14.111_audio-g3ab4335-ab31 | iter_3 | jank_frame | jankbench | std | -0.434760 | % | NaN | NaN
default-jankbench_edit_text | 4.14.111_audio-g3ab4335-ab31 | iter_3 | jank_frame | jankbench | count | 0.879121 | % | NaN | NaN
default-jankbench_edit_text | 4.14.111_audio-g3ab4335-ab31 | iter_3 | total_duration | jankbench | median | -11.820807 | % | NaN | NaN
default-jankbench_edit_text | 4.14.111_audio-g3ab4335-ab31 | iter_3 | total_duration | jankbench | mean | -10.863575 | % | NaN | NaN
default-jankbench_edit_text | 4.14.111_audio-g3ab4335-ab31 | iter_3 | total_duration | jankbench | std | 4.720619 | % | NaN | NaN
default-jankbench_edit_text | 4.14.111_audio-g3ab4335-ab31 | iter_3 | total_duration | jankbench | count | 0.879121 | % | NaN | NaN

File: lisa/wa.py
Comment (douglas-raillard-arm) [inline lisa/wa.py:562]: nit: s/WACollector/:class`lisa.stats.Stats`/ as the format is described by that class.

File: lisa/wa.py
Comment (douglas-raillard-arm) [inline lisa/wa.py:581]: it's a bit confusing as `wa` is a module which does not have any `get_collector` function. Would be better to stay closer to the other docstring example:
```
        wa_output = WAOutput('wa/output/path')
        df = wa_output['jankbench'].df
        ...
```

File: lisa/wa.py
Comment (douglas-raillard-arm) [inline lisa/wa.py:593]: _ARTIFACT_METRIC_SUFFIX_UNIT can be removed as it's not used anywhere

File: lisa/wa.py
Comment (douglas-raillard-arm) [inline lisa/wa.py:603]: would probably be a good idea to close the connection once we are done. The best way to do it would be to use:
https://docs.python.org/3/library/contextlib.html#contextlib.closing

```
from contextlib import closing
with closing(sqlite3.connect(path)) as con:
    raw_df = pd.read_sql_query("SELECT total_duration, jank_frame, name, _id as frame_id from ui_results", con) 

df = raw_df.melt(id_vars=['frame_id'], value_vars=['total_duration', 'jank_frame'])
return df 
```

**Review by douglas-raillard-arm — CHANGES_REQUESTED** [review]
Some small comments on lisa/wa.py otherwise LGTM.

Let's split the PR into 2:
1. lisa/wa.py with the new WAJankbenchCollector class that can be rebased on master and merged quickly once the comments are addressed
2. the example notebooks that might need more co-refining along plotting capabilities of lisa.stats.Stats

**Comment by credp** [comment]
Just going to test it again before pushing just the collector change back - I'll make a new PR for the notebooks.

File: lisa/wa.py
Comment (credp) [inline lisa/wa.py:20]: I see we already import the whole of contextlib below, so I'll remove this and access it with a path

**Comment by credp** [comment]
The notebooks are at https://github.com/credp/lisa/tree/jankbench_notebooks I'll make a new PR for them.

**Pull request merged**
